lib/client.js

```diff
@@ -124,10 +124,13 @@ function Client (broker, conn, req) {
         if (clientSub && (clientSub.qos || 0) < packet.qos) {
           packet.qos = clientSub.qos
         }
-        packet.writeCallback = cb
         if (that.clean || packet.retain) {
+          packet.writeCallback = cb
           writeQoS(null, that, packet)
         } else {
+          packet.writeCallback = () => {
+            packet.setDupIfNecessary(broker.persistence, that, cb)
+          }
```

Comment:
This will trigger two `outgoingUpdate` one just after the other, I think we could do this in only one instead

Reply:
I'm not sure about it. Isn't the order of the operations crucial here? Like:
1. Do the original update that was there before my changes.
2. Actually write the message to the stream.
3. Update the `dup` flag.

If something goes wrong in point 2. - eg. the stream is no longer available, then I suppose we don't want the `dup` flag updated.

Reply:
yeah you right by looking at this again seems there are no alternatives. From a code side it seems ok to me I would also like to know @mcollina and @gnought thoughts